### PR TITLE
Add ToolException that a tool can throw.

### PR DESCRIPTION
**Pull request by xming521**
# Add ToolException that a tool can throw
This is an optional exception that tool throws when execution error occurs.
When this exception is thrown, the agent will not stop working,but will handle the exception according to the handle_tool_error variable of the tool,and the processing result will be returned to the agent as observation,and printed in pink on the console.It can be used like this:
```python 
from langchain.schema import ToolException
from langchain import LLMMathChain, SerpAPIWrapper, OpenAI
from langchain.agents import AgentType, initialize_agent
from langchain.chat_models import ChatOpenAI
from langchain.tools import BaseTool, StructuredTool, Tool, tool
from langchain.chat_models import ChatOpenAI

llm = ChatOpenAI(temperature=0)
llm_math_chain = LLMMathChain(llm=llm, verbose=True)

class Error_tool:
    def run(self, s: str):
        raise ToolException('The current search tool is not available.')
    
def handle_tool_error(error) -> str:
    return "The following errors occurred during tool execution:"+str(error)

search_tool1 = Error_tool()
search_tool2 = SerpAPIWrapper()
tools = [
    Tool.from_function(
        func=search_tool1.run,
        name="Search_tool1",
        description="useful for when you need to answer questions about current events.You should give priority to using it.",
        handle_tool_error=handle_tool_error,
    ),
    Tool.from_function(
        func=search_tool2.run,
        name="Search_tool2",
        description="useful for when you need to answer questions about current events",
        return_direct=True,
    )
]
agent = initialize_agent(tools, llm, agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION, verbose=True,
                         handle_tool_errors=handle_tool_error)
agent.run("Who is Leo DiCaprio's girlfriend? What is her current age raised to the 0.43 power?")
```
image

## Who can review?
- @vowelparrot

### CODE REVIEW

File: langchain/tools/base.py [hunk 1]
```diff
@@ -251,8 +255,25 @@ def run(
...
             )
         except (Exception, KeyboardInterrupt) as e:
-            run_manager.on_tool_error(e)
-            raise e
+            if isinstance(e, ToolException):
+                if isinstance(self.handle_tool_error, bool):
+                    if self.handle_tool_error:
+                        observation = "Tool execution error"
+                    else:
+                        raise e
```
Comment (hwchase17) [inline langchain/tools/base.py:263]: should also on_tool_error here

Comment (xming521) [reply langchain/tools/base.py:264]: I added on_tool_error here.

**Comment by vowelparrot** [comment]
This looks good! We want to have similar behavior in `arun` - would you be able to mirror what you've implemented there?

**Comment by dev2049** [comment]
could we add an example notebook as well

**Comment by xming521** [comment]
@dev2049 @vowelparrot  Added documentation, supported the arun function, and made some adjustments.

**Comment by xming521** [comment]
@dev2049 Can you merge my code, please? Thank you.

**Review by vowelparrot — APPROVED** [review]

**Comment by dev2049** [comment]
thanks @xming521!

**Pull request merged**